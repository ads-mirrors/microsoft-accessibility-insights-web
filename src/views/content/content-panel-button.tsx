// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { Button } from '@fluentui/react-components';
import { FluentUIV9Icon } from 'common/icons/fluentui-v9-icons';
import * as React from 'react';
import { ContentActionMessageCreator } from '../../common/message-creators/content-action-message-creator';
import { NamedFC } from '../../common/react/named-fc';
import { ContentProvider, ContentReference } from './content-page';

export type ContentPanelButtonDeps = {
    contentProvider: ContentProvider;
    contentActionMessageCreator: ContentActionMessageCreator;
};

export type ContentPanelButtonProps = {
    contentTitle: string;
    deps: ContentPanelButtonDeps;
    reference: ContentReference | undefined;
    iconName?: string;
};

export const ContentPanelButton = NamedFC<ContentPanelButtonProps>(
    'ContentPanelButton',
    ({ contentTitle, deps, reference, children, iconName }) => {
        const { contentProvider, contentActionMessageCreator } = deps;

        if (!reference) {
            return null;
        }

        const contentPath = contentProvider.pathFromReference(reference);
        if (contentPath == null) {
            return null;
        }

        const onClick = ev =>
            contentActionMessageCreator?.openContentPanel(ev, contentPath, contentTitle);

        return (
            <Button
                className="info"
                appearance="transparent"
                icon={iconName && <FluentUIV9Icon iconName={iconName} />}
                onClick={onClick}
                aria-label={'info and examples'}
            >
                {children}
            </Button>
        );
    },
);
