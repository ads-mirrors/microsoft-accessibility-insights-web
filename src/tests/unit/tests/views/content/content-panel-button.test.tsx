// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { Button } from '@fluentui/react-components';
import { render } from '@testing-library/react';
import * as React from 'react';
import {
    expectMockedComponentPropsToMatchSnapshots,
    getMockComponentClassPropsForCall,
    mockReactComponents,
} from 'tests/unit/mock-helpers/mock-module-helpers';
import { ContentPage } from 'views/content/content-page';
import { ContentPanelButton } from 'views/content/content-panel-button';
import { ContentActionMessageCreator } from '../../../../../common/message-creators/content-action-message-creator';

jest.mock('@fluentui/react-components');
jest.mock('../../../../../common/message-creators/content-action-message-creator');

describe('ContentPanelButton', () => {
    mockReactComponents([Button]);
    const content = {
        for: {
            testing: ContentPage.create(() => 'CONTENT FOR TESTING' as any),
        },
    };
    const contentTitle = 'TITLE FOR TESTING';

    const deps = {
        contentProvider: ContentPage.provider(content),
        contentActionMessageCreator: {} as any as ContentActionMessageCreator,
    };

    it('renders from content', () => {
        const renderResult = render(
            <ContentPanelButton
                deps={deps}
                reference={content.for.testing}
                iconName="info"
                contentTitle={contentTitle}
            >
                TEXT
            </ContentPanelButton>,
        );

        expect(renderResult.asFragment()).toMatchSnapshot();
        expectMockedComponentPropsToMatchSnapshots([Button]);
    });

    it('renders from path', () => {
        const renderResult = render(
            <ContentPanelButton
                deps={deps}
                reference={'for/testing'}
                iconName="iconName"
                contentTitle={contentTitle}
            >
                TEXT
            </ContentPanelButton>,
        );

        expect(renderResult.asFragment()).toMatchSnapshot();
        expectMockedComponentPropsToMatchSnapshots([Button]);
    });

    it('renders without iconName', () => {
        const renderResult = render(
            <ContentPanelButton deps={deps} reference={'for/testing'} contentTitle={contentTitle}>
                TEXT
            </ContentPanelButton>,
        );

        expect(renderResult.asFragment()).toMatchSnapshot();
        expectMockedComponentPropsToMatchSnapshots([Button]);
    });

    it('renders with reference is null', () => {
        const renderResult = render(
            <ContentPanelButton deps={deps} reference={''} contentTitle={contentTitle}>
                TEXT
            </ContentPanelButton>,
        );
        expect(renderResult.asFragment()).toMatchSnapshot();
        expectMockedComponentPropsToMatchSnapshots([Button]);
    });

    it('button onClick', () => {
        const deps: any = {
            contentProvider: ContentPage.provider(content),
            contentActionMessageCreator: {
                openContentPanel: jest.fn(),
            },
        };
        render(
            <ContentPanelButton deps={deps} reference={'for/testing'} contentTitle={contentTitle}>
                TEXT
            </ContentPanelButton>,
        );

        getMockComponentClassPropsForCall(Button).onClick();
        expect(deps.contentActionMessageCreator.openContentPanel).toHaveBeenCalled();
    });
});
