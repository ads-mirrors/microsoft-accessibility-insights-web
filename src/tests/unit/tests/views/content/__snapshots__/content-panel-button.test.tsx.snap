// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContentPanelButton renders from content 1`] = `
<DocumentFragment>
  <mock-button
    appearance="transparent"
    aria-label="info and examples"
    classname="info"
    icon="[object Object]"
  >
    TEXT
  </mock-button>
</DocumentFragment>
`;

exports[`ContentPanelButton renders from content: Button props 1`] = `
{
  "appearance": "transparent",
  "aria-label": "info and examples",
  "children": "TEXT",
  "className": "info",
  "icon": <FluentUIV9Icon
    iconName="info"
  />,
  "onClick": [Function],
}
`;

exports[`ContentPanelButton renders from path 1`] = `
<DocumentFragment>
  <mock-button
    appearance="transparent"
    aria-label="info and examples"
    classname="info"
    icon="[object Object]"
  >
    TEXT
  </mock-button>
</DocumentFragment>
`;

exports[`ContentPanelButton renders from path: Button props 1`] = `
{
  "appearance": "transparent",
  "aria-label": "info and examples",
  "children": "TEXT",
  "className": "info",
  "icon": <FluentUIV9Icon
    iconName="iconName"
  />,
  "onClick": [Function],
}
`;

exports[`ContentPanelButton renders with reference is null 1`] = `<DocumentFragment />`;

exports[`ContentPanelButton renders with reference is null: Button props 1`] = `undefined`;

exports[`ContentPanelButton renders without iconName 1`] = `
<DocumentFragment>
  <mock-button
    appearance="transparent"
    aria-label="info and examples"
    classname="info"
  >
    TEXT
  </mock-button>
</DocumentFragment>
`;

exports[`ContentPanelButton renders without iconName: Button props 1`] = `
{
  "appearance": "transparent",
  "aria-label": "info and examples",
  "children": "TEXT",
  "className": "info",
  "icon": undefined,
  "onClick": [Function],
}
`;
